tempesta_fw/sched/tfw_sched_rr.c

```diff
@@ -101,42 +104,73 @@ tfw_sched_rr_add_conn(TfwSrvGroup *sg, TfwServer *srv, TfwConnection *conn)
 }
 
 /**
+ * Returns the next connection for the server.
+ * Dead connections are skipped.
+ */
+static inline TfwConnection *
+sched_rr_get_conn(TfwRrSrv *srv_cl)
+{
+	size_t c;
+
+	for (c = 0; c < srv_cl->conn_n; ++c) {
+		size_t idx = atomic64_inc_return(&srv_cl->rr_counter)
+				% srv_cl->conn_n;
+		TfwConnection *conn = srv_cl->conns[idx];
+
+		if (tfw_connection_get_if_nfo(conn))
+			return conn;
+	}
+	return NULL;
+}
+
+/**
  * On each subsequent call the function returns the next server in the group.
  * Parallel connections to the same server are also rotated in the
  * round-robin manner.
  * Dead connections and servers w/o live connections are skipped.
  */
 static TfwConnection *
-tfw_sched_rr_get_srv_conn(TfwMsg *msg, TfwSrvGroup *sg)
+tfw_sched_rr_sg_get_conn(TfwMsg *msg, TfwSrvGroup *sg)
 {
-	int c, s, i;
-	TfwConnection *conn;
+	size_t s;
 	TfwRrSrvList *sl = sg->sched_data;
-	TfwRrSrv *srv_cl;
 
 	BUG_ON(!sl);
 
 	for (s = 0; s < sl->srv_n; ++s) {
-		i = atomic64_inc_return(&sl->rr_counter) % sl->srv_n;
-		srv_cl = &sl->srvs[i];
-		for (c = 0; c < srv_cl->conn_n; ++c) {
-			i = atomic64_inc_return(&srv_cl->rr_counter)
-			    % srv_cl->conn_n;
-			conn = srv_cl->conns[i];
-			if (tfw_connection_get_if_nfo(conn))
-				return conn;
-		}
+		size_t idx = atomic64_inc_return(&sl->rr_counter) % sl->srv_n;
+		TfwConnection *conn = sched_rr_get_conn(&sl->srvs[idx]);
+
+		if (conn)
+			return conn;
 	}
 	return NULL;
 }
 
+/**
+ * Returns live connection to the server @srv.
+ * Parallel connections to the server are rotated in the round-robin manner.
+ */
+static TfwConnection *
+tfw_sched_rr_srv_get_conn(TfwMsg *msg, TfwServer *srv)
+{
+	TfwRrSrv *srv_cl = srv->sched_data;
+
+	/* For @srv without connections srv_cl will be NULL */
+	if (!srv_cl)
+		return NULL;
```

Comment:
I don't see a reason for removing unit tests for schedules about scheduling for a server group without connections. I will add `unlikely` here as a compromise.